src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,681 @@
+			                       video_encoder.frame) < 0) {
+				LOG_ERR("FFmpeg: Failed to send video frame");
+			}
+			send_packets_to_muxer(video_encoder.codec_context,
+			                      MuxerVideoStreamIndex,
+			                      muxer.queue);
+			image->free();
+		}
+
+		// Queue has been stopped. Flush the encoder by sending nullptr
+		// frame.
+		assert(!video_encoder.queue.IsRunning());
+		avcodec_send_frame(video_encoder.codec_context, nullptr);
+		send_packets_to_muxer(video_encoder.codec_context,
+		                      MuxerVideoStreamIndex,
+		                      muxer.queue);
+
+		lock.lock();
+		video_encoder.is_working = false;
+		lock.unlock();
+		waiter.notify_all();
+	}
+}
+
+void FFmpegEncoder::mux()
+{
+	for (;;) {
+		std::unique_lock<std::mutex> lock(mutex);
+		while (!muxer.is_initalised && !worker_threads_should_quit) {
+			waiter.wait(lock);
+		}
+		if (worker_threads_should_quit) {
+			return;
+		}
+		muxer.is_working = true;
+		lock.unlock();
+
+		while (auto optional = muxer.queue.Dequeue()) {
+			AVPacket* packet = *optional;
+			AVRational encoder_time_base;
+			if (packet->stream_index == MuxerVideoStreamIndex) {
+				encoder_time_base = video_encoder.codec_context->time_base;
+			} else {
+				encoder_time_base = audio_encoder.codec_context->time_base;
+			}
+			av_packet_rescale_ts(packet,
+			                     encoder_time_base,
+			                     muxer.format_context
+			                             ->streams[packet->stream_index]
+			                             ->time_base);
+			if (av_interleaved_write_frame(muxer.format_context,
+			                               packet) < 0) {
+				LOG_ERR("FFmpeg: Muxer failed to write frame");
+			}
+			av_packet_free(&packet);
+		}
+
+		// Pass nullptr to drain the muxer's internal buffer
+		av_interleaved_write_frame(muxer.format_context, nullptr);
+		av_write_trailer(muxer.format_context);
+		lock.lock();
+		muxer.is_working = false;
+		lock.unlock();
+		waiter.notify_all();
+	}
+}
+
+void FFmpegEncoder::encode_audio()
+{
+	std::vector<int16_t> audio_data;
+
+	for (;;) {
+		std::unique_lock<std::mutex> lock(mutex);
+		while (!audio_encoder.is_initalised && !worker_threads_should_quit) {
+			waiter.wait(lock);
+		}
+		if (worker_threads_should_quit) {
+			return;
+		}
+		audio_encoder.is_working = true;
+		lock.unlock();
+
+		// Number of samples the AVFrame has been allocated to hold
+		const int sample_capacity = audio_encoder.frame->nb_samples;
+
+		// 2 channels, 1 frame per channel
+		const size_t frame_capacity = sample_capacity * 2;
+
+		for (;;) {
+			audio_encoder.queue.BulkDequeue(audio_data, frame_capacity);
+			if (audio_data.empty()) {
+				assert(!audio_encoder.queue.IsRunning());
+				break;
+			}
+			if (av_frame_make_writable(audio_encoder.frame) < 0) {
+				LOG_ERR("FFmpeg: Failed to make audio frame writable");
+				continue;
+			}
+			const int received_samples = static_cast<int>(audio_data.size()) / 2;
```

Comment:
I agree with making samples vs frames clear.  This is samples.  The Audio RWQueue (which gets read into `audio_data` here) is in `int16_t` frames.  2 frames make 1 sample for stereo sound.  Right?  Or am I confusing the two?

Reply:
Ok, so here the `/ 2` is because we're dividing a size expressed in bytes by our sample size (2 bytes for 16-bit audio). That should be made clearer; these casual divide/multiply by 2 things can be very misleading and a source for bugs for the next guy.

Reply:
I was getting samples and frames mixed up in my head.  Changed these to the correct definitions.